jgrapht-core/src/main/java/org/jgrapht/alg/CyclicTransitiveReduction.java

```diff
@@ -0,0 +1,285 @@
+          scComponent.addEdge(sourceVertex, targetVertex);
+        }
+      }
+      else {
+        GraphPath<V, E> graphPath = new DirectedHamiltonianCycle<V, E>().getTour(scComponent);
+        assert graphPath != null;
+        cycle = graphPath.getVertexList();
+        // First and last vertices are identical -> remove one of them
+        cycle.remove(0);
+
+        for (E edge : sccEdgesCopy) {
+          int sourceIndex = cycle.indexOf(scComponent.getEdgeSource(edge));
+          assert sourceIndex >= 0;
+          int targetIndex = cycle.indexOf(scComponent.getEdgeTarget(edge));
+          assert targetIndex >= 0;
+          int indexDelta = Math.abs(sourceIndex - targetIndex);
+          if (indexDelta != 1 && indexDelta + 1 != cycle.size())
+            scComponent.removeEdge(edge);
+        }
+      }
+    }
+  }
+
+  /**
+   * Re-expand a condensed graph previously created by {@link #condenseGraph(Graph)} after possible clean-ups by
+   * {@link #pruneCondensedGraph(Graph)}) back into the original directed graph, also cleaning up  redundant edges
+   * between strongly connected components during the process.
+   *
+   * @param condensedGraph condensed graph to be re-expanded
+   * @param directedGraph target graph to be expanded into
+   */
+  protected void expandCondensedGraph(Graph<Graph<V, E>, DefaultEdge> condensedGraph, Graph<V, E> directedGraph) {
+    // Per condensed graph edge, find one existing edge in the original graph to represent
+    // the corresponding connection between two SCCs
+    Set<E> sccLinks = new HashSet<>();
+    condensedGraph.edgeSet()
+      .forEach(edge -> {
+        Set<V> sourceVertices = condensedGraph.getEdgeSource(edge).vertexSet();
+        Set<V> targetVertices = condensedGraph.getEdgeTarget(edge).vertexSet();
+        E sccLink = allowSyntheticEdges
+          ? createSyntheticSCCLink(sourceVertices, targetVertices)
+          : findSCCLink(sourceVertices, targetVertices);
+        assert sccLink != null;
+        sccLinks.add(sccLink);
+      });
+
+    // Remove all edges from the original graph except for the SCC links
+    new HashSet<E>(directedGraph.edgeSet()).stream()
+      .filter(edge -> !sccLinks.contains(edge))
+      .forEach(directedGraph::removeEdge);
+
+    // Re-add all edges within SCCs
+    condensedGraph.vertexSet().forEach(scComponent ->
+      scComponent.edgeSet().forEach(edge ->
+        directedGraph.addEdge(scComponent.getEdgeSource(edge), scComponent.getEdgeTarget(edge))
+      )
+    );
+  }
+
+  /**
+   * Find an existing edge in a directed graph, linking two SCCs (strongly connected components) within that graph
+   *
+   * @param sourceVertices source SCC vertex set
+   * @param targetVertices target SCC vertex set
+   * @return the first edge found which links any source vertex to any target vertex
+   */
+  private E findSCCLink(Set<V> sourceVertices, Set<V> targetVertices) {
+    for (V sourceVertex : sourceVertices) {
```

Comment:
Feel free to do so. I know next to nothing about the internals of JGraphT, I even was a first time user before creating this PR. Possibly there is a lot of existing functionality I implemented, but I did not find it to be accessible where I needed it. I wanted to avoid refactoring the core product. BTW, things like an adjacency matrix, adjacency lists, edges/vertices in a defined order or accessing edge properties "source/target" would also be nice to be accessible in order to avoid unnecessary transformations, but they are not AFAIK. So I just used the existing API.

Reply:
@jsichi, any ideas or suggestions here? This is an open review issue, but affects another class. Can I actively do something in order to resolve it?

Reply:
I've created #1010 to track this.  You can just add a link to that for now without implementing it (to avoid scope creep).  Later, after this current PR gets merged, we can move on to that as a followup optimization in a separate PR.

Reply:
Links added